### ISSUE TITLE
Partial constraints for GCPs

### ISSUE THREAD

**Issue by stephanmantler** [mapillary/OpenSfM#302]
Is it posssible to provide partial constraints for GCPs?

I am monitoring glacier change with monthly measurements, and currently need to align each flight manually due to the lack of precise ground control points. Because of the way the landscape changes, it would be relatively easy to establish repeatable, precise lat/lon coordinates for certain features even without precise geolocation, but altitude may change significantly between measurement runs. 

Conversely there is a glacial lake with known altitude but it would be almost impossible to use lat/lon out there. So I was wondering if the GCP control point format could be extended to allow something like

    NAN NAN <geo_z> <img_x> <img_y> <image_name>
    <geo_x> <geo_y> NAN <img_x> <img_y> <image_name>

I poked a bit around the source code but couldn't quite figure out how the cost function of gcp observations is evaluated.

**Comment by paulinus** [comment]
tl;dr: not possible with the way we handle GPS right now. Possible with a different implementation.

As currently implemented, GCP are assumed to be points with known 3D position. These points are projected to the images and compared with their observed position there.

Such procedure can only be made if all 3 coords of the GPC are known.

In reality, though, we often don't know the exact position. Sometimes the estimate might be uncertain. Other times, as in the case you mention, only some axes are observed.

A better way to handle such situations would be to estimate the 3D location of the GPC instead of assuming that they are known. This way the estimated position could be constrained to be close to different types of measurements including the ones where only some axes are observed.

**Comment by yjmenezes** [comment]
@stephanmantler 

i think a Least Square Ajustment with GCP weight would be a solution.
If you like, take a look into this example.
To run, 
[LeastSquareAdjustment.pdf](https://github.com/mapillary/OpenSfM/files/1928661/LeastSquareAdjustment.pdf)
 please install GNU Octave and unpack under /tmp/

[odm_affine3d_2017-04-16.zip](https://github.com/mapillary/OpenSfM/files/1928658/odm_affine3d_2017-04-16.zip)
